geomstats/stratified_geometry/wald_space.py

```diff
@@ -0,0 +1,1024 @@
+            "Information geometry for phylogenetic trees"
+            Journal of Mathematical Biology, 82(3):19, February 2021a.
+            https://doi.org/10.1007/s00285-021-01553-x.
+[Lueg21]_   Lueg, J., M. K. Garba, T. M. W. Nye, S. F. Huckemann.
+            "Wald Space for Phylogenetic Trees."
+            Geometric Science of Information, Lecture Notes in Computer Science,
+            pages 710–717, Cham, 2021.
+            https://doi.org/10.1007/978-3-030-80209-7_76.
+"""
+
+
+import functools
+import itertools as it
+
+import numpy as np
+
+import geomstats.backend as gs
+import geomstats.geometry.spd_matrices as spd
+import geomstats.stratified_geometry.stratified_spaces
+from geomstats.stratified_geometry.stratified_spaces import Point, PointSet
+
+
+@functools.total_ordering
+class Split:
+    r"""Class for two-set partitions of sets.
+
+    Two-set partitions of a smaller subset of :math:`\{0,...,n-1\}` are also allowed,
+    where :math:`n` is a positive integer, which is not passed as an argument as it is
+    nowhere needed.
+
+    The parameters ``part1`` and ``part2`` are assigned to the attributes ``self.part1``
+    and ``self.part2`` in a unique sorted way: the one that contains the smallest
+    minimal value is assigned to ``self.part1`` for consistency.
+
+    Parameters
+    ----------
+    part1 : iterable
+        The first part of the split, an iterable that is a subset of
+        :math:`\{0,\dots,n-1\}`. It may be empty, but must have empty intersection with
+        ``part2``.
+    part2 : iterable
+        The second part of the split, an iterable that is a subset of
+        :math:`\{0,\dots,n-1\}`. It may be empty, but must have empty intersection with
+        ``part1``.
+    """
+
+    def __init__(self, part1, part2):
+        part1, part2 = tuple(sorted(list(part1))), tuple(sorted(list(part2)))
+        if set(part1) & set(part2):
+            raise ValueError(
+                f"A split consists of disjoint sets, those are not: {part1}, {part2}."
+            )
+        if part1 and part2:
+            self.part1 = part1 if part1[0] < part2[0] else part2
+            self.part2 = part2 if part1[0] < part2[0] else part1
+        elif not part1:
+            self.part1 = part2
+            self.part2 = ()
+        elif not part2:
+            self.part1 = part1
+            self.part2 = ()
+        else:
+            self.part1 = ()
+            self.part2 = ()
+
+    def restrict_to(self, subset: set):
+        r"""Return the restriction of a split to a subset.
+
+        Parameters
+        ----------
+        subset : set
+            The subset that the split is restricted to.
+
+        Returns
+        -------
+        restr_split : Split
+            The restricted split, if the split is :math:`A\vert B`, then the split
+            restricted to the subset :math:`C` is :math:`A\cap C\vert B\cap C`.
+        """
+        return Split(
+            part1=tuple(set(self.part1) & subset),
```

Comment:
Is there any need to "setify" `self.part1` and `self.part2`? We can ensure they are sets from the beginning (or force them to be, if the case). Probably does not have performance impact, but is an unnecessary operation.

This repeats several times.